Share same existsQuery impl throughout mappers

javanna opened the pull request:
Most of our field types have the same implementation for their `existsQuery` method which relies on doc_values if present, otherwise it queries norms if available or uses a term query against the _field_names meta field. This standard implementation is repeated in many different mappers.

There are field types that only query doc_values, because they always have them, and field types that always query _field_names, because they never have norms nor doc_values. We could apply the same standard logic to all of these field types as `MappedFieldType` has the knowledge about what data structures are available.

This commit introduces a standard implementation that does the right thing depending on the data structure that is available. With that only field types that require a different behaviour need to override the existsQuery method.

At the same time, this no longer forces subclasses to override `existsQuery`, which could be forgotten when needed. To address this we introduced a new test method in `MapperTestCase` that verifies the `existsQuery` being generated and its consistency with the available data structures.

elasticmachine commented:
Pinging @elastic/es-search (:Search/Mapping)

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:
```diff
@@ -312,7 +314,21 @@ public Query regexpQuery(String value, int flags, int maxDeterminizedStates, @Nu
+    public Query existsQuery(QueryShardContext context) {
+        if (hasDocValues()) {
+            return new DocValuesFieldExistsQuery(name());
+        } else if (omitNorms()) {
```
I believe it should be `omitNorms() || indexOptions() != IndexOptions.NONE`

javanna replied:
I gave up on this, leaving the norms based impl where they are (it's only 3 of them). The standard impl only switches between doc_values and field_names

javanna commented:
failures are due to fields that are indexed with points for which `omitNorms == false && indexOptions() != NONE`, yet norms are not available as the inverted index is empty, so the norms exists lucene query never matches any document.

jpountz reviewed (commented):
One thing I'm worried about with this PR is that if we implemented a virtual field like `constant_keyword` today after your PR is merged, it would be easy to forget to override `existsQuery`. Should we keep the method abstract in MappedFieldType and introduce a base MappedFieldType impl for concrete fields that would have this implementation that is common to keywords, numbers, ...?

javanna committed:
Share same existsQuery impl throughout mappers

Most of our field types have the same implementation for their existsQuery method which relies on doc_values if present, otherwise it queries norms if available or uses a term query against the _field_names meta field. This standard implementation is repeated in many different mappers.

There are field types that only query doc_values, because they always have them, and field types that always query _field_names, because they never have norms nor doc_values. We could apply the same standard logic to all of these field types.

This commit introduces a standard implementation that does the right thing depending on the data structure that is available. With that only field types that require a different behaviour need to override the existsQuery method.

One advantage of having existsQuery implemented in each mapper may have been that each mapper also decides whether to create or not the _field_names field. Though _field_names should only be created whenever doc_values are turned off: this assumption is now enforced through an assertion.

javanna committed:
Merge branch 'master' into enhancement/mappers_same_exists_query

javanna committed:
address KeywordFieldTypeTests#testexistsQuery

javanna committed:
Try out a new base class for the shared existsQuery

javanna commented:
@romseygeek I brought this back to life, and thanks to all your refactoring, the base class implementation works now. We longer have issues with omit norms and index options that we used to have. Would you mind having a look?

> One thing I'm worried about with this PR is that if we implemented a virtual field like constant_keyword today after your PR is merged, it would be easy to forget to override existsQuery. Should we keep the method abstract in MappedFieldType and introduce a base MappedFieldType impl for concrete fields that would have this implementation that is common to keywords, numbers, ...?

@jpountz it is a valid concern, though many many classes share this common implementation, and the others look more like outliers to me where special treatment is needed. I gave this a try in the last commit I pushed, and it is not so bad, but I would have liked to make the method final in the separate base class. That is not possible due to the existing class hierarchy from `SimpleMappedFieldType` and corresponding subclasses. I think the simplest way would be without this additional base class and just have to remember to override `existsQuery` for the minority of field type where the "concrete" impl does not apply. Curious to hear your thoughts based on this attempt.

romseygeek commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapper.java:
```diff
@@ -126,6 +126,7 @@ public DocValueFormat docValueFormat(String format, ZoneId timeZone) {
 
         @Override
         public Query existsQuery(QueryShardContext context) {
+            //TODO how can this work if this field never has doc_values?
```
The vector is stored in a BinaryDocValuesField, so I think this is fine.

javanna replied:
ok but then shouldn't it pass doc_values set to true to the super class?

romseygeek replied:
It looks as though `hasDocValues()` is used exclusively for a) indexing (which should go away once everything is properly parametrized) and b) building `exist` queries; it's separate from `isAggregatable` which is used for field caps. So yes, the vector fields should probably set `doc_values` to true in the constructor.

javanna replied:
I opened #62631 to fix this and add tests for it

romseygeek commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/SparseVectorFieldMapper.java:
```diff
@@ -104,6 +104,7 @@ public DocValueFormat docValueFormat(String format, ZoneId timeZone) {
 
         @Override
         public Query existsQuery(QueryShardContext context) {
+            //TODO how can this work if this field never has doc_values?
```
Same as above

javanna replied:
this is a bit tricky because the field is no longer supported, fielddataBuilder was removed yet there are doc_values in the index. I will figure it out

javanna replied:
This is addressed by #62646

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/ConcreteMappedFieldType.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.util.Map;
+
+public abstract class ConcreteMappedFieldType extends MappedFieldType {
```
Can you add some javadoc just to explain why we need this intermediate class?

javanna replied:
I am still hoping that we don't need this class, do you have thoughts? :) I will add javadocs if we decide to keep it.

romseygeek replied:
I'm easy either way, I think.  It might be that adding a base class test would also act as a reminder to implementers?  Something like:
- build an existsQuery and parse a document with the relevant field
- if its a NormsFieldExistsQuery then check the doc contains a field with norms enabled
- if its a DocValuesFieldExistsQuery then check the doc contains a doc values field
- if its a TermQuery then check the doc contains a `field_names` field
- otherwise we expect a MatchAllDocsQuery or MatchNoDocsQuery

Then mapper implementations that don't actually index anything will throw an error if they don't return a MatchAllDocsQuery or MatchNoDocsQuery

javanna replied:
I like the idea!

javanna replied:
I pushed an update that has a new test added to `MapperTestCase` for this. I think you'll have ideas around how to make it better, it still has some rough edges:

- binary field is special because its exists query don't work unless the field is stored or has doc_values, we end up querying field_names even when it's not there, maybe we should change the implementation to return MatchNoDocsQuery in that case.
- we are testing only the default settings for each field type. It would be nice to test also the different combinations around omitNorms and hasDocValues but not all field types allow to set them which would require more specific methods to be overridden etc.

romseygeek reviewed (approved):
LGTM, one nit around javadoc but otherwise I think it's good to go.

I think it would be nice as a follow up to add a test to MapperTestCase that checks that the exists query for a mapper lines up with the contents of a parsed document - I can look into that, as it may not be entirely straightforward.

javanna committed:
Merge branch 'master' into enhancement/mappers_same_exists_query

javanna committed:
test existsQuery behaviour

romseygeek commented on test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java:
```diff
@@ -53,6 +62,50 @@
+    public final void testExistsQuery() throws IOException {
+        MapperService mapperService = createMapperService(fieldMapping(this::minimalMapping));
+        ParsedDocument doc = mapperService.documentMapper().parse(source(this::field));
+        withLuceneIndex(mapperService, iw -> iw.addDocument(doc.rootDoc()), reader -> {
```
I don't think you need to index the document, you can just look at the fields in the parsed doc?

javanna replied:
good point, thanks for raising this, will fix

romseygeek commented on test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java:
```diff
@@ -53,6 +62,50 @@
+        assertParseMinimalWarnings();
+    }
+
+    protected void assertExistsQuery(MappedFieldType fieldType, QueryShardContext queryShardContext, IndexReader reader) {
```
I think you can change this to just take `MapperService`.  Then we have the default `testExistsQuery` impl which calls this with a minimal mapping; and mappers which have norms or doc_values settings can then have their own tests that call this with the relevant configuration.

javanna replied:
this is implemented now, thanks for the tip.

romseygeek commented on test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java:
```diff
@@ -53,6 +62,50 @@
     protected abstract void minimalMapping(XContentBuilder b) throws IOException;
 
+    protected void field(XContentBuilder builder) throws IOException {
+        builder.field("field");
```
I think the only mappers that need to override this are ones that do not extend ConcreteFieldType, so maybe we can add a check in the test and take a different branch if we know we need to do something different here?

javanna replied:
I am overriding this at the moment  in constant keyword and runtime field mapper. there are concrete field types (for instance the ones that throw when calling existsQuery) that don't subclass ConcreteFieldType, which is partly why I don't like to have that base class. 

This method is really for the field types that should not write the field in the docs at all. Can you expand a bit on what you are suggesting, I am not sure I am getting it?

romseygeek replied:
> there are concrete field types (for instance the ones that throw when calling existsQuery) that don't subclass ConcreteFieldType

ah ok, never mind then, I think I misunderstood what was going on

romseygeek reviewed (changes requested):
Thanks @javanna, I have a couple of suggestions.

javanna committed:
expand tests

javanna committed:
Merge branch 'master' into enhancement/mappers_same_exists_query

javanna committed:
iter

javanna committed:
add javadocs

javanna commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -285,6 +285,8 @@ public Query existsQuery(QueryShardContext context) {
                 return new DocValuesFieldExistsQuery(name());
             } else {
+                //TODO this field differs from all other fields as it never goes into _field_names, so the default impl would not work
+                //though changing this is problematic for existing indices that don't have this field in _field_names?
```
Hopefully the test that I added does this.

javanna committed:
spotless

romseygeek commented on modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentIdFieldMapper.java:
```diff
@@ -123,11 +121,6 @@ public Object valueForDisplay(Object value) {
-        public Query existsQuery(QueryShardContext context) {
-            return new DocValuesFieldExistsQuery(name());
-        }
     }
```
It's annoying that neither this nor ParentJoinFieldMapper actually have tests for existsQuery, but they'll get them once we convert their test classes to MapperTestCase and it works by inspection now.

javanna replied:
agreed

romseygeek replied:
I opened https://github.com/elastic/elasticsearch/issues/62774 to convert all the remaining test cases

javanna committed:
remove ConcreteMappedFieldType.java

javanna committed:
checkstyle

romseygeek reviewed (approved):
LGTM!

javanna committed:
fix test

javanna committed:
Merge branch 'master' into enhancement/mappers_same_exists_query

javanna committed:
Merge branch 'master' into enhancement/mappers_same_exists_query

javanna commented:
thanks for all the help @romseygeek !
